[popover] Implement [popover=auto] and light dismiss behavior

nt1m opened the pull request:
#### 725f6435785d7cb7f5d84c6e11c8639e521db6db
<pre>
[popover] Implement [popover=auto] and light dismiss behavior
<a href="https://bugs.webkit.org/show_bug.cgi?id=252080">https://bugs.webkit.org/show_bug.cgi?id=252080</a>
rdar://105613012

Reviewed by Darin Adler.

<a href="https://html.spec.whatwg.org/#the-popover-attribute">https://html.spec.whatwg.org/#the-popover-attribute</a>

<a href="https://html.spec.whatwg.org/#popover-light-dismiss">https://html.spec.whatwg.org/#popover-light-dismiss</a>

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/hide-other-popover-side-effects-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/light-dismiss-event-ordering-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-anchor-nesting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-stacking-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::topmostAutoPopover const):
(WebCore::Document::hideAllPopoversUntil):
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::defaultEventHandler):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::topmostPopoverAncestor):
(WebCore::HTMLElement::queuePopoverToggleEventTask):
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::defaultKeyboardEventHandler):

Canonical link: <a href="https://commits.webkit.org/261093@main">https://commits.webkit.org/261093@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd9d1f191e8edf87beaa0b68cc20bcd24ef777f7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110577 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19657 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/1933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114532 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/21082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/10789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/102804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116319 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/21082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/102804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/21082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/102804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12310 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/12872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/10789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/18236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7687 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14754 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8933,6 +8933,66 @@ HTMLDialogElement* Document::activeModalDialog() const
+    for (auto& element : makeReversedRange(m_topLayerElements)) {
+        if (auto* candidate = downcast<HTMLElement>(element.ptr()); candidate->popoverState() == PopoverState::Auto)
+            return candidate;
+    }
```
If there is no guarantee these are HTML elements, then we need a dynamic downcast:
```suggestion
    for (auto& element : makeReversedRange(m_topLayerElements)) {
        if (auto* candidate = dynamicDowncast<HTMLElement>(element.get()); candidate->popoverState() == PopoverState::Auto)
            return candidate;
    }
```

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8933,6 +8933,66 @@ HTMLDialogElement* Document::activeModalDialog() const
+            continue;
+        if (element.ptr() == endpoint)
+            break;
+        popoversToHide.append(downcast<HTMLElement>(element.get()));
```
Seems like an unnecessarily awkward idiom.
```suggestion
        if (!is<HTMLElement>(element))
            continue;
        auto& popover = downcast<HTMLElement>(element.get());
        if (popover.popoverState() != PopoverState::Auto || &popover == endpoint)
            break;
        popoversToHide.append(popover);
```

nt1m replied:
The suggested code isn't quite correct, since we want to keep iterating if the popoverState isn't auto.

`auto& popover = downcast<HTMLElement>(element.get());` also isn't quite correct, because a html element in top layer could also be a dialog element, or a fullscreen element, not necessarily a popover (so the popover state, as the spec calls it, would be none).

nt1m replied:
Changed it to this, which avoids the downcasting twice:

```
    Vector<Ref<HTMLElement>> popoversToHide;
    for (auto& item : makeReversedRange(m_topLayerElements)) {
        if (!is<HTMLElement>(item))
            continue;
        auto& element = downcast<HTMLElement>(item.get());
        if (element.popoverState() == PopoverState::Auto) {
            if (&element == endpoint)
                break;
            popoversToHide.append(element);
        }
    }
```

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8933,6 +8933,66 @@ HTMLDialogElement* Document::activeModalDialog() const
+                continue;
+            if (element->popoverData()->visibilityState() != PopoverVisibilityState::Showing)
+                continue;
+            return element;
```
```suggestion
            auto& element = downcast<HTMLElement>(*node);
            if (element.popoverState() != PopoverState::Auto)
                continue;
            if (element.popoverData()->visibilityState() != PopoverVisibilityState::Showing)
                continue;
            return &element;
```

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8933,6 +8933,66 @@ HTMLDialogElement* Document::activeModalDialog() const
+                continue;
+            return element;
+        }
+        // TODO: Handle popovertarget attributes.
```
WebKit uses FIXME.
```suggestion
        // FIXME: Handle popovertarget attributes.
```

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8933,6 +8933,66 @@ HTMLDialogElement* Document::activeModalDialog() const
+    }
+
+    ASSERT(event.type() == eventNames().pointerupEvent);
+    if (m_popoverPointerDownTarget == popoverToAvoidHiding.get())
```
The == should work without get().
```suggestion
    if (m_popoverPointerDownTarget == popoverToAvoidHiding)
```

nt1m replied:
It seems like this doesn't work, WeakPtr can't compare against RefPtr, this is why I have to use `.get()`

darinadler commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2499,6 +2500,9 @@ void Node::defaultEventHandler(Event& event)
 #endif
+    } else if (eventType == eventNames.pointerdownEvent || eventType == eventNames.pointerupEvent) {
+        if (is<PointerEvent>(event))
+            document().handlePopoverLightDismiss(event);
```
If we are going to do a type check here, then should `handlePopoverLightDismiss` take a `PointerEvent&`?

darinadler commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8933,6 +8933,66 @@ HTMLDialogElement* Document::activeModalDialog() const
+        return;
+
+    RefPtr popoverToAvoidHiding = [&]() -> HTMLElement* {
+        for (Node* node = downcast<Node>(event.target()); node; node = node->parentInComposedTree()) {
```
Can we something more like `lineageOfType<HTMLElement>` here?

nt1m replied:
The only thing preventing from using `lineageOfType` is really the fact the `event.target()` is a `Node`, not an `Element`.

darinadler replied:
I thought that function takes a `Node` argument. I remember Antti did it that way.

darinadler replied:
I see now, `ancestorsOfType` takes `Node`, but `lineageOfType` takes `Element`. This might be a design mistake. Should talk to Antti about it, so we have the best pattern for long-term use.

nt1m replied:
I think it might have been intentional, since `lineageOfType` includes itself which has to be an `Element` type in order for `lineageOfType<ElementType>()` to make sense (you can't do `ancestorsOfType<Node>()` or `lineageOfType<Node>()`.

But I can check with the relevant people to see if this can be improved in some way.

darinadler replied:
Yes, of course it was intentional as far as that goes. But if what you have is a node that _might_ be an element, forcing you to do the type check explicitly, if it might be an element of the type that `lineage` is looking for, may not be the best design.

If we end up writing our own loop, by definition this function template has failed at its mission

nt1m committed:
[popover] Implement [popover=auto] and light dismiss behavior
https://bugs.webkit.org/show_bug.cgi?id=252080
rdar://105613012

Reviewed by Darin Adler.

https://html.spec.whatwg.org/#the-popover-attribute

https://html.spec.whatwg.org/#popover-light-dismiss

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/hide-other-popover-side-effects-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/light-dismiss-event-ordering-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-anchor-nesting-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-stacking-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-target-element-disabled-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-attribute-basic-expected.txt.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-focus-expected.txt.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt: Copied from LayoutTests/imported/w3c/web-platform-tests/html/semantics/popovers/popover-light-dismiss-expected.txt.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::topmostAutoPopover const):
(WebCore::Document::hideAllPopoversUntil):
(WebCore::Document::handlePopoverLightDismiss):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::defaultEventHandler):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::topmostPopoverAncestor):
(WebCore::HTMLElement::queuePopoverToggleEventTask):
(WebCore::HTMLElement::showPopover):
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::defaultKeyboardEventHandler):

Canonical link: https://commits.webkit.org/261093@main
